File: src/lib/notebooks.ts
import officialRegistry from '@/content/notebooks.json';
import type { RepositoryBundle } from './repository';
import { kindPath } from './repository';

import apiMcpSource from '../../notebooks/getting-started/super-ii-api-and-mcp.ipynb?raw';
import datasetSource from '../../notebooks/repositories/create-and-verify-a-dataset.ipynb?raw';
import evaluationSource from '../../notebooks/evaluation/reproducible-model-evaluation.ipynb?raw';

export type NotebookTextOutput = {
  mime_type: 'text/plain';
  text: string;
};

export type NotebookJsonOutput = {
  mime_type: 'application/json';
  value: unknown;
};

export type NotebookImageOutput = {
  mime_type: 'image/png' | 'image/jpeg' | 'image/webp';
  data: string;
  size_bytes: number;
};

export type NotebookOutput =
  | { type: 'stream'; name: 'stdout' | 'stderr'; text: string }
  | { type: 'error'; ename: string; evalue: string; traceback: string[] }
  | {
      type: 'display_data' | 'execute_result';
      execution_count?: number | null;
      data: Array<NotebookTextOutput | NotebookJsonOutput | NotebookImageOutput>;
    };

export type NotebookCell = {
  index: number;
  id?: string;
  type: 'markdown' | 'code' | 'raw';
  source: string;
  execution_count?: number | null;
  outputs?: NotebookOutput[];
};

export type NotebookDocument = {
  version: 1;
  path: string;
  size_bytes: number;
  nbformat: 4;
  nbformat_minor: number;
  kernel: {
    name: string | null;
    display_name: string | null;
    language: string | null;
    language_version: string | null;
  };
  cell_count: number;
  cells: NotebookCell[];
  omitted_outputs: number;
  safety: {
    static_only: true;
    code_executed: false;
    repository_code_imported: false;
    raw_html_rendered: false;
    javascript_rendered: false;
    svg_rendered: false;
    widgets_rendered: false;
    attachments_rendered: false;
  };
};

export type OfficialNotebookMeta = (typeof officialRegistry)[number];

const officialSources: Record<string, string> = {
  'notebooks/getting-started/super-ii-api-and-mcp.ipynb': apiMcpSource,
  'notebooks/repositories/create-and-verify-a-dataset.ipynb': datasetSource,
  'notebooks/evaluation/reproducible-model-evaluation.ipynb': evaluationSource,
};

function object(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function joinedText(value: unknown): string | null {
  if (typeof value === 'string') return value;
  if (Array.isArray(value) && value.every((part) => typeof part === 'string')) {
    return value.join('');
  }
  return null;
}

function boundedString(value: unknown, maximum: number): string | null {
  return typeof value === 'string' && value.length <= maximum ? value : null;
}

function integer(value: unknown, fallback = 0): number {
  return Number.isInteger(value) && Number(value) >= 0 ? Number(value) : fallback;
}

function kernelFromMetadata(metadata: Record<string, unknown> | null): NotebookDocument['kernel'] {
  const kernelspec = object(metadata?.kernelspec);
  const language = object(metadata?.language_info);
  return {
    name: boundedString(kernelspec?.name, 200),
    display_name: boundedString(kernelspec?.display_name, 200),
    language: boundedString(language?.name, 200),
    language_version: boundedString(language?.version, 200),
  };
}

function normalizedKernel(value: unknown): NotebookDocument['kernel'] {
  const kernel = object(value);
  return {
    name: boundedString(kernel?.name, 200),
    display_name: boundedString(kernel?.display_name, 200),
    language: boundedString(kernel?.language, 200),
    language_version: boundedString(kernel?.language_version, 200),
  };
}

function safeOutput(value: unknown): NotebookOutput | null {
  const entry = object(value);
  if (!entry) return null;
  if (entry.type === 'stream') {
    const text = boundedString(entry.text, 250_000);
    if (text === null) return null;
    return { type: 'stream', name: entry.name === 'stderr' ? 'stderr' : 'stdout', text };
  }
  if (entry.type === 'error') {
    const ename = boundedString(entry.ename, 200);
    const evalue = boundedString(entry.evalue, 1_000);
    const traceback = Array.isArray(entry.traceback)
      ? entry.traceback.filter((line): line is string => typeof line === 'string').slice(0, 100)
      : [];
    if (
      ename === null
      || evalue === null
      || traceback.some((line) => line.length > 250_000)
      || traceback.reduce((total, line) => total + line.length, 0) > 250_000
    ) {
      return null;
    }
    return { type: 'error', ename, evalue, traceback };
  }
  if (entry.type !== 'display_data' && entry.type !== 'execute_result') return null;
  const data: Array<NotebookTextOutput | NotebookJsonOutput | NotebookImageOutput> = [];
  for (const raw of Array.isArray(entry.data) ? entry.data.slice(0, 8) : []) {
    const item = object(raw);
    if (!item) continue;
    if (item.mime_type === 'text/plain') {
      const text = boundedString(item.text, 250_000);
      if (text !== null) data.push({ mime_type: 'text/plain', text });
    } else if (item.mime_type === 'application/json') {
      data.push({ mime_type: 'application/json', value: item.value });
    } else if (
      item.mime_type === 'image/png'
      || item.mime_type === 'image/jpeg'
      || item.mime_type === 'image/webp'
    ) {
      const encoded = boundedString(item.data, 7_000_000);
      const size = integer(item.size_bytes);
      if (encoded !== null && size <= 5 * 1024 * 1024 && /^[A-Za-z0-9+/]*={0,2}$/.test(encoded)) {
        data.push({ mime_type: item.mime_type, data: encoded, size_bytes: size });
      }
    }
  }
  if (!data.length) return null;
  return {
    type: entry.type,
    execution_count: entry.type === 'execute_result'
      ? entry.execution_count === null ? null : integer(entry.execution_count)
      : undefined,
    data,
  };
}

export function normalizeRepositoryNotebook(value: unknown): NotebookDocument | null {
  const notebook = object(value);
  const safety = object(notebook?.safety);
  if (
    !notebook
    || notebook.version !== 1
    || notebook.nbformat !== 4
    || safety?.static_only !== true
    || safety.code_executed !== false
    || safety.raw_html_rendered !== false
    || safety.javascript_rendered !== false
  ) return null;
  const path = boundedString(notebook.path, 1_024);
  if (!path || path.includes('..') || path.startsWith('/')) return null;
  const rawCells = Array.isArray(notebook.cells) ? notebook.cells.slice(0, 2_000) : null;
  if (!rawCells) return null;
  const cells: NotebookCell[] = [];
  for (const [position, rawCell] of rawCells.entries()) {
    const cell = object(rawCell);
    const type = cell?.type;
    const source = boundedString(cell?.source, 250_000);
    if (!cell || !['markdown', 'code', 'raw'].includes(String(type)) || source === null) return null;
    const normalized: NotebookCell = {
      index: integer(cell.index, position),
      type: type as NotebookCell['type'],
      source,
    };
    const id = boundedString(cell.id, 64);
    if (id) normalized.id = id;
    if (type === 'code') {
      normalized.execution_count = cell.execution_count === null
        ? null
        : integer(cell.execution_count);
      normalized.outputs = (Array.isArray(cell.outputs) ? cell.outputs : [])
        .map(safeOutput)
        .filter((output): output is NotebookOutput => output !== null);
    }
    cells.push(normalized);
  }
  return {
    version: 1,
    path,
    size_bytes: integer(notebook.size_bytes),
    nbformat: 4,
    nbformat_minor: integer(notebook.nbformat_minor),
    kernel: normalizedKernel(notebook.kernel),
    cell_count: cells.length,
    cells,
    omitted_outputs: integer(notebook.omitted_outputs),
    safety: {
      static_only: true,
      code_executed: false,
      repository_code_imported: false,
      raw_html_rendered: false,
      javascript_rendered: false,
      svg_rendered: false,
      widgets_rendered: false,
      attachments_rendered: false,
    },
  };
}

function normalizeOfficialNotebook(source: string, path: string): NotebookDocument | null {
  let notebook: Record<string, unknown> | null = null;
  try {
    notebook = object(JSON.parse(source));
  } catch {
    return null;
  }
  if (
    !notebook
    || notebook.nbformat !== 4
    || !Array.isArray(notebook.cells)
    || notebook.cells.length > 2_000
  ) return null;
  const cells: NotebookCell[] = [];
  let omittedOutputs = 0;
  for (const [index, rawCell] of notebook.cells.entries()) {
    const cell = object(rawCell);
    const type = cell?.cell_type;
    const cellSource = joinedText(cell?.source);
    if (
      !cell
      || !['markdown', 'code', 'raw'].includes(String(type))
      || cellSource === null
      || cellSource.length > 250_000
    ) return null;
    const normalized: NotebookCell = {
      index,
      type: type as NotebookCell['type'],
      source: cellSource,
    };
    const id = boundedString(cell.id, 64);
    if (id) normalized.id = id;
    if (type === 'code') {
      normalized.execution_count = null;
      normalized.outputs = [];
      omittedOutputs += Array.isArray(cell.outputs) ? cell.outputs.length : 0;
    }
    cells.push(normalized);
  }
  return {
    version: 1,
    path,
    size_bytes: new TextEncoder().encode(source).byteLength,
    nbformat: 4,
    nbformat_minor: integer(notebook.nbformat_minor),
    kernel: kernelFromMetadata(object(notebook.metadata)),
    cell_count: cells.length,
    cells,
    omitted_outputs: omittedOutputs,
    safety: {
      static_only: true,
      code_executed: false,
      repository_code_imported: false,
      raw_html_rendered: false,
      javascript_rendered: false,
      svg_rendered: false,
      widgets_rendered: false,
      attachments_rendered: false,
    },
  };
}

export function repositoryNotebooks(repository: RepositoryBundle): NotebookDocument[] {
  const analysis = repository.analyses.find(
    (entry) => entry.analysis_type === 'notebook' && entry.status === 'passed',
  );
  const notebooks = object(analysis?.result)?.notebooks;
  return (Array.isArray(notebooks) ? notebooks : [])
    .map(normalizeRepositoryNotebook)
    .filter((notebook): notebook is NotebookDocument => notebook !== null);
}

export function repositoryNotebookHref(repository: RepositoryBundle, path: string): string {
  const encodedPath = path.split('/').map(encodeURIComponent).join('/');
  return `/${kindPath(repository.kind)}/${encodeURIComponent(repository.owner_handle)}/${encodeURIComponent(repository.slug)}/notebooks/${encodedPath}`;
}

export const officialNotebooks = officialRegistry;

export function getOfficialNotebook(slug: string | undefined): {
  meta: OfficialNotebookMeta;
  notebook: NotebookDocument;
  githubUrl: string;
  colabUrl: string;
  downloadUrl: string;
} | null {
  const meta = officialRegistry.find((item) => item.slug === slug);
  if (!meta) return null;
  const source = officialSources[meta.path];
  const notebook = source ? normalizeOfficialNotebook(source, meta.path) : null;
  if (!notebook) return null;
  const githubUrl = `https://github.com/smavgs/super-ii/blob/main/${meta.path}`;
  return {
    meta,
    notebook,
    githubUrl,
    colabUrl: `https://colab.research.google.com/github/smavgs/super-ii/blob/main/${meta.path}`,
    downloadUrl: `https://raw.githubusercontent.com/smavgs/super-ii/main/${meta.path}`,
  };
}

File: runtime/src/superii_runtime/inspectors/notebooks.py
from __future__ import annotations

import base64
import binascii
import json
from pathlib import Path
from typing import Any

import nbformat
from nbformat import ValidationError

MAX_NOTEBOOK_BYTES = 25 * 1024 * 1024
MAX_NOTEBOOKS = 32
MAX_CELLS = 2_000
MAX_CELL_SOURCE_CHARS = 250_000
MAX_TOTAL_SOURCE_CHARS = 5_000_000
MAX_OUTPUTS_PER_CELL = 100
MAX_TEXT_OUTPUT_CHARS = 250_000
MAX_TOTAL_OUTPUT_CHARS = 8_000_000
MAX_IMAGE_BYTES = 5 * 1024 * 1024
MAX_TOTAL_IMAGE_BYTES = 16 * 1024 * 1024
SAFE_IMAGE_MIME_TYPES = {"image/jpeg", "image/png", "image/webp"}


class NotebookInspectionError(ValueError):
    """Raised when a notebook cannot safely enter the static-view contract."""


def _text(value: Any) -> str:
    if isinstance(value, str):
        return value
    if isinstance(value, list) and all(isinstance(part, str) for part in value):
        return "".join(value)
    raise NotebookInspectionError("notebook text fields must contain strings")


def _bounded_text(value: Any, *, limit: int, label: str) -> str:
    text = _text(value)
    if len(text) > limit:
        raise NotebookInspectionError(f"{label} exceeds the static-view limit")
    return text


def _metadata_text(value: Any, *, limit: int = 200) -> str | None:
    if not isinstance(value, str):
        return None
    clean = value.strip()
    return clean[:limit] if clean else None


def _safe_json(value: Any) -> Any:
    try:
        encoded = json.dumps(value, ensure_ascii=False, allow_nan=False, separators=(",", ":"))
    except (TypeError, ValueError) as error:
        raise NotebookInspectionError("notebook JSON output is not serializable") from error
    if len(encoded) > MAX_TEXT_OUTPUT_CHARS:
        raise NotebookInspectionError("notebook JSON output exceeds the static-view limit")
    return json.loads(encoded)


def _safe_image(mime_type: str, value: Any) -> dict[str, Any]:
    encoded = _bounded_text(value, limit=MAX_IMAGE_BYTES * 2, label=mime_type).replace("\n", "")
    try:
        decoded = base64.b64decode(encoded, validate=True)
    except (binascii.Error, ValueError) as error:
        raise NotebookInspectionError(f"{mime_type} output is not valid base64") from error
    if len(decoded) > MAX_IMAGE_BYTES:
        raise NotebookInspectionError(f"{mime_type} output exceeds the static-view limit")
    return {"mime_type": mime_type, "data": encoded, "size_bytes": len(decoded)}


def _safe_output(output: Any) -> tuple[dict[str, Any] | None, int, int]:
    output_type = output.get("output_type")
    if output_type == "stream":
        text = _bounded_text(
            output.get("text", ""), limit=MAX_TEXT_OUTPUT_CHARS, label="stream output"
        )
        return (
            {
                "type": "stream",
                "name": "stderr" if output.get("name") == "stderr" else "stdout",
                "text": text,
            },
            len(text),
            0,
        )
    if output_type == "error":
        traceback = [
            _bounded_text(line, limit=MAX_TEXT_OUTPUT_CHARS, label="traceback line")
            for line in output.get("traceback", [])[:100]
        ]
        ename = _metadata_text(output.get("ename")) or "Error"
        evalue = _metadata_text(output.get("evalue"), limit=1_000) or ""
        text_size = len(ename) + len(evalue) + sum(len(line) for line in traceback)
        if text_size > MAX_TEXT_OUTPUT_CHARS:
            raise NotebookInspectionError("error output exceeds the static-view limit")
        return (
            {"type": "error", "ename": ename, "evalue": evalue, "traceback": traceback},
            text_size,
            0,
        )
    if output_type not in {"display_data", "execute_result"}:
        return None, 0, 0

    data = output.get("data")
    if not isinstance(data, dict):
        raise NotebookInspectionError("rich notebook output must contain a MIME bundle")
    safe_data: list[dict[str, Any]] = []
    text_size = 0
    image_size = 0
    if "text/plain" in data:
        text = _bounded_text(
            data["text/plain"], limit=MAX_TEXT_OUTPUT_CHARS, label="text/plain output"
        )
        safe_data.append({"mime_type": "text/plain", "text": text})
        text_size += len(text)
    if "application/json" in data:
        value = _safe_json(data["application/json"])
        safe_data.append({"mime_type": "application/json", "value": value})
        text_size += len(json.dumps(value, ensure_ascii=False))
    for mime_type in sorted(SAFE_IMAGE_MIME_TYPES):
        if mime_type in data:
            image = _safe_image(mime_type, data[mime_type])
            safe_data.append(image)
            image_size += int(image["size_bytes"])
    if not safe_data:
        return None, 0, 0
    safe: dict[str, Any] = {"type": output_type, "data": safe_data}
    if output_type == "execute_result":
        count = output.get("execution_count")
        safe["execution_count"] = count if isinstance(count, int) and count >= 0 else None
    return safe, text_size, image_size


def inspect_notebook(path: Path, root: Path) -> dict[str, Any]:
    resolved = path.resolve(strict=True)
    size_bytes = resolved.stat().st_size
    if size_bytes > MAX_NOTEBOOK_BYTES:
        raise NotebookInspectionError(f"{path.name} exceeds the 25 MiB notebook limit")
    try:
        raw = json.loads(resolved.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError) as error:
        raise NotebookInspectionError(f"{path.name} is not valid UTF-8 notebook JSON") from error
    if not isinstance(raw, dict) or raw.get("nbformat") != 4:
        raise NotebookInspectionError(f"{path.name} must use Jupyter nbformat 4")
    try:
        notebook = nbformat.from_dict(raw)
        nbformat.validate(notebook)
    except (AttributeError, TypeError, ValidationError, ValueError) as error:
        raise NotebookInspectionError(f"{path.name} failed nbformat validation") from error

    cells = notebook.get("cells", [])
    if len(cells) > MAX_CELLS:
        raise NotebookInspectionError(f"{path.name} exceeds the cell-count limit")
    safe_cells: list[dict[str, Any]] = []
    total_source_chars = 0
    total_output_chars = 0
    total_image_bytes = 0
    omitted_outputs = 0
    for index, cell in enumerate(cells):
        source = _bounded_text(
            cell.get("source", ""), limit=MAX_CELL_SOURCE_CHARS, label=f"cell {index} source"
        )
        total_source_chars += len(source)
        if total_source_chars > MAX_TOTAL_SOURCE_CHARS:
            raise NotebookInspectionError(f"{path.name} exceeds the total source limit")
        cell_type = cell.get("cell_type")
        safe_cell: dict[str, Any] = {
            "index": index,
            "type": cell_type,
            "source": source,
        }
        cell_id = _metadata_text(cell.get("id"), limit=64)
        if cell_id:
            safe_cell["id"] = cell_id
        if cell_type == "code":
            count = cell.get("execution_count")
            safe_cell["execution_count"] = count if isinstance(count, int) and count >= 0 else None
            outputs = cell.get("outputs", [])
            if len(outputs) > MAX_OUTPUTS_PER_CELL:
                raise NotebookInspectionError(f"cell {index} exceeds the output-count limit")
            safe_outputs: list[dict[str, Any]] = []
            for output in outputs:
                safe_output, text_chars, image_bytes = _safe_output(output)
                if safe_output is None:
                    omitted_outputs += 1
                    continue
                total_output_chars += text_chars
                total_image_bytes += image_bytes
                if total_output_chars > MAX_TOTAL_OUTPUT_CHARS:
                    raise NotebookInspectionError(f"{path.name} exceeds the text-output limit")
                if total_image_bytes > MAX_TOTAL_IMAGE_BYTES:
                    raise NotebookInspectionError(f"{path.name} exceeds the image-output limit")
                safe_outputs.append(safe_output)
            safe_cell["outputs"] = safe_outputs
        safe_cells.append(safe_cell)

    metadata = notebook.get("metadata", {})
    kernelspec = metadata.get("kernelspec", {}) if isinstance(metadata, dict) else {}
    language = metadata.get("language_info", {}) if isinstance(metadata, dict) else {}
    return {
        "version": 1,
        "path": resolved.relative_to(root.resolve(strict=True)).as_posix(),
        "size_bytes": size_bytes,
        "nbformat": 4,
        "nbformat_minor": int(notebook.get("nbformat_minor", 0)),
        "kernel": {
            "name": _metadata_text(kernelspec.get("name"))
            if isinstance(kernelspec, dict)
            else None,
            "display_name": (
                _metadata_text(kernelspec.get("display_name"))
                if isinstance(kernelspec, dict)
                else None
            ),
            "language": (
                _metadata_text(language.get("name")) if isinstance(language, dict) else None
            ),
            "language_version": (
                _metadata_text(language.get("version")) if isinstance(language, dict) else None
            ),
        },
        "cell_count": len(safe_cells),
        "cells": safe_cells,
        "omitted_outputs": omitted_outputs,
        "safety": {
            "static_only": True,
            "code_executed": False,
            "repository_code_imported": False,
            "raw_html_rendered": False,
            "javascript_rendered": False,
            "svg_rendered": False,
            "widgets_rendered": False,
            "attachments_rendered": False,
        },
    }


def inspect_notebooks(root: Path) -> dict[str, Any]:
    """Validate and normalize local notebooks without importing or executing their code."""

    resolved = root.resolve(strict=True)
    notebooks = sorted(resolved.rglob("*.ipynb"))
    if len(notebooks) > MAX_NOTEBOOKS:
        raise NotebookInspectionError("revision exceeds the notebook-count limit")
    inspected = [inspect_notebook(path, resolved) for path in notebooks]
    return {
        "version": 1,
        "notebook_count": len(inspected),
        "notebooks": inspected,
        "static_only": True,
        "code_executed": False,
    }
